### PR TITLE
Add Tacotron2 loss function

### PR DESCRIPTION
**Pull request by yangarbiter**
The main Tacotron2 model is in #1621. Since that PR is already too large, I'll put the Tacotron2 loss and it's tests in the example directory temporary. We should discuss more on where to put this loss function (likely the same file as Tacotron2's model).

### CODE REVIEW

**Review by vincentqb — COMMENTED** [review]
please move all the files related to the loss (e.g. `loss_function.py` and tests) in a subfolder `examples/pipeline_tacotron2/loss/`

File: examples/pipeline_tacotron2/loss_function.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        r"""Pass the input through the Tacotron2 loss.
+
+        Args:
+            model_outputs (tuple of three Tensors): The outputs of the
+                Tacotron2. These outputs should include three items:
+                (1) the predicted mel spectrogram before the postnet (mel_specgram) with shape (n_batch, n_mel, n_time),
+                (2) predicted mel spectrogram after the postnet (mel_specgram_postnet)
+                    with shape (n_batch, n_mel, n_time), and
+                (3) the stop token prediction (gate_out) with shape (n_batch).
+            targets (tuple of two Tensors): The ground truth mel spectrogram (n_batch, n_mel, n_time) and
```
Comment (vincentqb) [inline examples/pipeline_tacotron2/loss_function.py:53]: nit: we write `(batch, mel, time), see [here](https://github.com/pytorch/audio/blob/master/README.md#conventions)

File: examples/pipeline_tacotron2/loss_function.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# *****************************************************************************
```
[the remark this answers is not in the document]
Comment (yangarbiter) [reply examples/pipeline_tacotron2/loss_function.py:1]: According to the [internal document](https://docs.google.com/document/d/1gap2DWVKHCo_my8cRxQFiHgfz4-JzGO8nsr1XYHFz1o/edit?usp=sharing), it appears that we should implement it in examples folder first. We can then discuss whether we want to merge it into the core later on.
Please let me know how you think about it. :)
Thanks.

File: examples/pipeline_tacotron2/test_tacotron2_loss.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        temp_dir = os.path.join(self.get_base_temp_dir(), self.id())
+        path = os.path.join(temp_dir, *paths)
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        return path
+
+
+class Tacotron2LossTest(unittest.TestCase, TempDirMixin):
+
+    dtype = torch.float64
+    device = "cpu"
```
[the remark this answers is not in the document]
Comment (yangarbiter) [reply examples/pipeline_tacotron2/test_tacotron2_loss.py]: I've add GPU tests [here](https://github.com/pytorch/audio/pull/1625/commits/42b866c7486cbd4f2f40b1d9abaefac8e5dd1627#diff-cf2183a63ca7da8a49e0c8c875c779e97fe2248349d83bc68b3abea6ace9c6a7R116) and [here](https://github.com/pytorch/audio/pull/1625/commits/42b866c7486cbd4f2f40b1d9abaefac8e5dd1627#diff-cf2183a63ca7da8a49e0c8c875c779e97fe2248349d83bc68b3abea6ace9c6a7R190)

**Comment by yangarbiter** [comment]
Following @vincentqb's suggestion, I've also run  the code through [black](https://github.com/psf/black).

File: examples/pipeline_tacotron2/loss_function.py (hunk 1 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/loss_function.py]: Thanks.
Addressed [here](https://github.com/pytorch/audio/pull/1625/files#diff-446b770bff29efd0f406a17635f004c65a136ace539bd3d4eea923e14de8915dR56)

File: examples/pipeline_tacotron2/loss_function.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+from typing import Tuple
+
+from torch import nn, Tensor
+
+
+class Tacotron2Loss(nn.Module):
+    """Tacotron2 loss function adapted from:
+    https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/SpeechSynthesis/Tacotron2/tacotron2/loss_function.py
```
[the remark this answers is not in the document]
Comment (yangarbiter) [reply examples/pipeline_tacotron2/loss_function.py]: Thanks for pointing it out. I've addressed it [here](https://github.com/pytorch/audio/pull/1625/files#diff-446b770bff29efd0f406a17635f004c65a136ace539bd3d4eea923e14de8915dR48).

File: examples/pipeline_tacotron2/loss/test_tacotron2_loss.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    return (
+        mel_specgram,
+        mel_specgram_postnet,
+        gate_out,
+        truth_mel_specgram,
+        truth_gate_out,
+    )
+
+
+class Tacotron2LossTest(unittest.TestCase, TempDirMixin):
```
Comment (vincentqb) [inline examples/pipeline_tacotron2/loss/test_tacotron2_loss.py:77]: please follow the convention we have for testing cpu/gpu

File: examples/pipeline_tacotron2/loss/test_tacotron2_loss.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        )
+        gradgradcheck(
+            _fn,
+            (mel_specgram, mel_specgram_postnet, gate_out, truth_mel_specgram, truth_gate_out),
+            fast_mode=True,
+        )
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (vincentqb) [inline examples/pipeline_tacotron2/loss/test_tacotron2_loss.py:197]: please remove this as we do not have this in test files anymore

File: examples/pipeline_tacotron2/loss/test_tacotron2_loss.py (hunk 5 above)
Comment (vincentqb) [reply examples/pipeline_tacotron2/loss/test_tacotron2_loss.py:77]: please also add a test that validates the shape of the output.

File: examples/pipeline_tacotron2/loss/test_tacotron2_loss.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+    def get_temp_path(self, *paths):
+        temp_dir = os.path.join(self.get_base_temp_dir(), self.id())
+        path = os.path.join(temp_dir, *paths)
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        return path
+
+
+def _get_inputs(dtype, device):
+    n_mel, n_batch, max_mel_specgram_length = 3, 2, 4
```
Comment (vincentqb) [inline examples/pipeline_tacotron2/loss/test_tacotron2_loss.py:55]: these seem very small and are not representative of a typical tensor shape. please make `n_mel` and `_max_mel_specgram_length` larger, following examples from the other tests.

File: examples/pipeline_tacotron2/loss/test_tacotron2_loss.py (hunk 6 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/loss/test_tacotron2_loss.py]: Thanks for the suggestion, it is removed.

File: examples/pipeline_tacotron2/loss/test_tacotron2_loss.py (hunk 5 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/loss/test_tacotron2_loss.py]: Thanks for the suggestion.
The shape test is [here](https://github.com/pytorch/audio/pull/1625/files#diff-2405e8a4b292ae3cadb917a86765e3158d22489dc14f00e895c25bdf19604e22R63).
And the base class and subclasses is also addressed in the same [commit](https://github.com/pytorch/audio/pull/1625/commits/5b9942bc3f07e4b314a2a5e25a6c2b172379bce4).

**Review by vincentqb — APPROVED** [review]
LGTM. @carolineechen @mthrok -- do you have any other feedback?

**Review by mthrok — COMMENTED** [review]
How would you like to treat the test? We can run them as a a part of CI.
Also can you document how to run the test?

**Comment by yangarbiter** [comment]
> How would you like to treat the test? We can run them as a a part of CI.
Yes, I think it is a good idea to run them as part of CI. Could you direct me on how to do that properly?
Thanks.

> Also can you document how to run the test?
Sure, I'll add them.

**Comment by yangarbiter** [comment]
> > > How would you like to treat the test? We can run them as a a part of CI.
> > 
> > 
> > Yes, I think it is a good idea to run them as part of CI. Could you direct me on how to do that properly?
> > Thanks.
> > Should I edit the code [here](https://github.com/pytorch/audio/blob/master/.circleci/unittest/linux/scripts/run_test.sh#L26) by adding
> > ```shell
> > cd ../examples/pipeline_tacotron2/
> > pytest pytest "${args[@]}" .
> > cd ../../test
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I use the whole directory (pipeline_tacotron2) because there are going to be tests for text utilities too.
> > > Also can you document how to run the test?
> > 
> > 
> > Sure, I've added a short description at `examples/pipeline_tacotron2/README.md`
> 
> Ideally, we want to run the test in a separate job, but for a starter, this is good. You do not need `"${args[@]}"` as that is mostly intended for unit test. I would simply add `-v`. i.e. `pytest -v .`

Thanks for the comment.
The tests are add in `.circleci/unittest/linux/scripts/run_test.sh` and `.circleci/unittest/windows/scripts/run_test.sh`.

File: examples/pipeline_tacotron2/loss/tacotron2_loss_gpu_test.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import torch
+
+from .tacotron2_loss_impl import (
+    Tacotron2LossShapeTests,
+    Tacotron2LossTorchscriptTests,
+    Tacotron2LossGradcheckTests,
+)
+
+
+def skipIfNoCuda(test_item):
```
Comment (mthrok) [inline examples/pipeline_tacotron2/loss/tacotron2_loss_gpu_test.py:13]: ~Can you add `#TODO: Find a way to de-duplicate this utility and reuse the original one from "test/torchaudio_unittest"`~

~Looks like multiple utilities are duplicated, and this soon becomes unmanageable. Can you modify `PYTHONPATH` to include the `<repo_root>/test/` directory so that you can import?~

Actually, I remembered that `example` directories are exposed in unit tests. 

https://github.com/pytorch/audio/blob/c49db7393a096ff18725603f5132707c7a3bef83/test/torchaudio_unittest/example/__init__.py#L1-L8

Can you move the test implementations into [`test/torchaudio_unittest/example`](https://github.com/pytorch/audio/tree/master/test/torchaudio_unittest/example
) directory?

File: examples/pipeline_tacotron2/loss/__init__.py [hunk 9]
```diff
@@ -0,0 +1 @@
+from .loss_function import Tacotron2Loss
```
Comment (mthrok) [inline examples/pipeline_tacotron2/loss/__init__.py:1]: Please add a new line.

Comment (yangarbiter) [reply examples/pipeline_tacotron2/loss/__init__.py]: I've removed `examples/pipeline_tacotron2/loss` directory and moved everything to `examples/pipeline_tacotron2/loss.py` as the tests are moved to `test/torchaudio_unittest/example`.

File: .circleci/unittest/linux/scripts/run_test.sh [hunk 10]
```diff
@@ -24,4 +24,7 @@ declare -a args=(
 
 cd test
 pytest "${args[@]}" torchaudio_unittest
+cd ../examples/pipeline_tacotron2/
+pytest -v .
+cd ../../test
```
[the remark this answers is not in the document]
Comment (yangarbiter) [reply .circleci/unittest/linux/scripts/run_test.sh]: I've removed these as all the tests are moved to `test/torchaudio_unittest/example`.

File: examples/pipeline_tacotron2/loss/tacotron2_loss_gpu_test.py (hunk 8 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/loss/tacotron2_loss_gpu_test.py]: Good idea! I've moved the test implementations to  `test/torchaudio_unittest/example/tacotron2`.

**Review by mthrok — APPROVED** [review]
